compiler-rt/lib/scudo/standalone/secondary.h

```diff
@@ -334,9 +336,6 @@ class MapAllocatorCache {
       }
 
       insert(Entry);
```

Comment:
I had a look at the unit test failure and it looks like the DCHECK in `release()` fails. Looking at the code, it seems we may actually insert entries with time values of `0` at the from of the LRU list because of this code:
```
if (useMemoryTagging<Config>(Options)) {
      if (Interval == 0 && !SCUDO_FUCHSIA) {
        // Release the memory and make it inaccessible at the same time by
        // creating a new MAP_NOACCESS mapping on top of the existing mapping.
        // Fuchsia does not support replacing mappings by creating a new mapping
        // on top so we just do the two syscalls there.
        Entry.Time = 0;
        mapSecondary<Config>(Options, Entry.CommitBase, Entry.CommitSize,
                             Entry.CommitBase, MAP_NOACCESS, Entry.MemMap);
      } else {
        Entry.MemMap.setMemoryPermission(Entry.CommitBase, Entry.CommitSize,
                                         MAP_NOACCESS);
      }
    }
```
at the start of the `store()` function. Need to figure out if test failure occurs in an environment where MTE is enabled.

Reply:
The reason why I'm sure if MTE is causing the problem is that we fixed a [bug](https://github.com/llvm/llvm-project/pull/107927) which caused some problem on MTE enabled devices. If the qemu has MTE enabled, I suppose it would catch the bug as well but it didn't (To be fair, it has the chance that the tests were not good enough to catch it in the past). Although I tend to believe the problem is what we are discussing here, just need some time to double check.

About
```
Instead of inserting the released entry at the head of the list, we could simply insert it after LastUnreleasedEntry to preserve the LRU ordering.
```
I would prefer handling the `LastUnreleasedEntry` in store() otherwise you need to know if the pages are released in the insert() function which is supposed to only handle the list insertion. In addition, at the time we remove the `Time`, it implies that we may need to pass another argument to convey the pages status. So I'm leaning toward handing them in store() but I'm not sure if there's any other restrictions.

Reply:
My thinking was that the insert() and remove() functions should be responsible for maintaining all the metadata for the list. I don't see a problem with updating the value in store() unless insert() starts being used more than once, in which case we would need to update `LastUnreleasedEntry` outside of insert() every time we call it. I will note, however, that we already now whether or not the pages are released within insert() because insert() takes in a `CachedBlock::Entry` which will have its time value set to `0` if it has been released. This way we can handle the logic for inserting released entries directly in insert as well as updating `LastUnreleasedEntry`.

Reply:
The case I'm saying is the moment when we want to remove the time record in the `CacheEntry`. Which means we  may need another argument to tell the page status but anyway, it's not a problem here so we can do it as what you suggest. BTW, I think the issue was the cases mentioned above, would you prepare a reapply CL for this?

Reply:
Oh ok, I see. I have updated the pull request accordingly at #110391.